src/dns.rs

```diff
@@ -0,0 +1,217 @@
+    // List of addresses for this domain
+    ips: Vec<IpAddr>,
+
+    /// True when the ips list has ipv4 addresses
+    has_ipv4: bool,
+    /// True when the ips list has ipv6 addresses
+    has_ipv6: bool,
+
+    // Internal iterator pointer
+    iter: usize,
+    /// expiry time after epoch
+    expires: u64,
+}
+
+impl DnsEntry {
+    /// Instantiate a new domain name entry with set of ips
+    pub(crate) fn new(domain: &str, ips: Vec<&str>) -> DnsEntry {
+        let mut entry = DnsEntry {
+            domain: domain.to_string(),
+            ..Default::default()
+        };
+
+        for ip in ips {
+            if let Ok(ip) = IpAddr::from_str(ip) {
+                if ip.is_ipv4() {
+                    entry.has_ipv4 = true;
+                }
+                if ip.is_ipv6() {
+                    entry.has_ipv6 = true;
+                }
+                entry.ips.push(ip);
+            }
+        }
+
+        entry
+    }
+
+    /// Returns true if the dns entry has expired
+    pub fn expired(&self) -> bool {
+        self.expires
+            < SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs()
+    }
+
+    fn ipv4(&self) -> Vec<IpAddr> {
+        self.ips.iter().filter(|x| x.is_ipv4()).copied().collect()
+    }
+
+    fn ipv6(&self) -> Vec<IpAddr> {
+        self.ips.iter().filter(|x| x.is_ipv6()).copied().collect()
+    }
+}
+
+impl Iterator for DnsEntry {
+    type Item = IpAddr;
+
+    /// With next() you can simply iterate over all the ips in the list
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.iter >= self.ips.len() {
+            // reset iterator at the end
+            self.iter = 0;
+
+            return None;
+        }
+
+        let ip = self.ips[self.iter];
+        self.iter += 1;
+
+        Some(ip)
+    }
+}
+
+/// Type of DNS resolution
+#[derive(Display, Debug, PartialEq, Clone)]
+pub enum ResolveType {
+    /// Only resolve IPV4 addresses (A)
+    Ipv4,
+    /// Only resolve IPV6 addresses (AAAA)
+    Ipv6,
+    /// Resolve both IPV4 and IPV6 addresses
+    Both,
+}
+
+trait DnsResolver {
+    /// Resolves a domain name for a given resolver_type
+    fn resolve(&mut self, domain: &str, resolve_type: ResolveType) -> Result<DnsEntry>;
+    /// Announces the resolved dns entry for the domain to a resolver
+    fn announce(&mut self, domain: &str, entry: &DnsEntry);
+
+    // name for debugging purposes
+    fn name(&self) -> &'static str;
+}
+
+// impl fmt::Debug for dyn DnsResolver {
+//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+//         write!(f, "dns resolver")
+//     }
+// }
+
+trait DnsCache {
+    /// Flush all domains
+    fn flush_all(&mut self);
+    /// Flush a single domain
+    fn flush_entry(&mut self, domain: &str);
+}
+
+struct Dns {
+    resolvers: Vec<Box<dyn DnsResolver>>,
+}
+
+impl Dns {
+    pub fn new() -> Self {
+        let mut resolvers: Vec<Box<dyn DnsResolver>> = vec![];
```

Comment:
I would not worry about this too much. It's not used in any thight loops, and the first push will do an allocation for 4 elements anyway. 
So instead of with_capacity(3), you get a with_capacity(4), which is ok enough.